### ISSUE TITLE
Import rules from tosback.org

### ISSUE THREAD

**Issue by MattiSG** [ambanum/OpenTermsArchive#58]
As discussed this morning with @michielbdejong and @Ndpnt:

- Import rules from https://github.com/tosdr/tosback2 by converting them into service declarations.
- Validate each imported service declaration with `scripts/validation/validate.js`.
  - Drop (or fix) service declarations that fail validation.
- Convert Xpath selectors into CSS selectors.
  - If an Xpath expression is not convertible into a CSS selector, find a CSS selector manually.
  - If there are too many inconvertible Xpath selectors, add support for the `xpathContentSelector` key in service declarations, which would rely on [`document.evaluate`](https://developer.mozilla.org/en-US/docs/Web/API/Document/evaluate) supposedly exposed by JSdom.
- Add a metadata entry at the root of the service declaration: `"importedFrom": "https://github.com/tosdr/tosback2/blob/5acac7abb5e967cfafd124a5e275f98f6ecd423e/rules/4shared.com.xml"`. Make sure to use stable URLs (with a commit ID) and not a branch reference 🙂 Adding this key will require changing the [validation schema](https://github.com/ambanum/CGUs/blob/master/scripts/validation/service.schema.js).

**Comment by michielbdejong** [comment]
Script is done in scripts/tosback-import.js. I included both import from XML (https://tosback.org) and from postgres (https://edit.tosdr.org). Now running the script, this will increase our number of documents from 100 to 2000.
* some will be OK
* some will be valid according to our validation script, but will still give false positives in practice, will just have to try crawling them once an hour for a while and see where false positives show up
* some do not pass validation
* some don't have a valid document type assigned yet
* some can't be parsed (should find out why)
* some can't be fetched (should discard them)
* some are a document type that we don't want to crawl (should decide whether we can keep a single criterion for relevant document types across the projects)
* some are a in a language other than English (should make sure we crawl the English document instead)

**Comment by michielbdejong** [comment]
I'm working on this in https://github.com/ambanum/TOSBack-CGUs-bridge/issues/2

**Comment by michielbdejong** [comment]
Result from yesterday's run using all the latest bugfixes:
https://github.com/ambanum/CGUs/pull/173

**Comment by Ndpnt** [comment]
Just copy the comment in the PR #173, as it's more appropriate here:

> Even if I think you are already aware of this, I just want to clarify our needs for service importation:
> - For services which pass the validation, it's important to us they comply with service naming standards described in #106 
> - For services which are successfully converted (document type is supported in CGUs and xpath to css conversion succeed) but do not pass the validation, can you save the resulting declarations in a dedicated folder in the TOSBack-CGUs-bridge repository?
> - For services which are not successfully converted (document type is unsupported in CGUs or xpath to css conversion failed), can you save the resulting declarations (with the original document type and xpath selector in `select` key)  in another dedicated folder in the TOSBack-CGUs-bridge repository?

**Comment by michielbdejong** [comment]
Thanks!
> For services which pass the validation, it's important to us they comply with service naming standards described in #106

Yes, this is done in https://github.com/ambanum/TOSBack-CGUs-bridge/commit/e25205b71535fa7c83f2eff0e2600a0ef0b42ee1
Let me know if you still see any incorrect service names in the results of tonight's run, tomorrow.

> For services which are successfully converted (document type is supported in CGUs and xpath to css conversion succeed) but do not pass the validation, can you save the resulting declarations in a dedicated folder in the TOSBack-CGUs-bridge repository?

Sure! That would not be too much work. So in https://github.com/ambanum/TOSBack-CGUs-bridge/issues/8#issuecomment-707712039 that would be the bottom 4 categories:
```sh
 39	inconsistent
       6	too stort
       5	not fetchable
       7	selector not found
```

> For services which are not successfully converted (document type is unsupported in CGUs or xpath to css conversion failed), can you save the resulting declarations (with the original document type and xpath selector in select key) in another dedicated folder in the TOSBack-CGUs-bridge repository?

Sure, that would be:
```sh
     12	Same type
     107	Unsupported type
```

Will try to do that tonight or tomorrow.

**Comment by MattiSG** [comment]
Opened in #184, partly delivered in #195, left to do documented in #196.

**Comment by MattiSG** [comment]
This track of work is over. We have reached a stage where we have way more services than we can even handle, so this import has certainly been a success 😄 All declarations have been moved to https://github.com/OpenTermsArchive/services-all. We migrated the declarations to that other repository, in order to clarify the distinction between the Open Terms Archive engine and the documents themselves, which can be maintained by different entities 🙂

If there is ever a need to import all the leftovers from ToSBack, the matching branch could be resuscitated from #196.

Thanks everyone for your contributions 🙂 👍
